lib/bgpstream_bgpdump.c

```diff
@@ -0,0 +1,234 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors:
+ *   Mingwei Zhang
+ *   Alistair King
+ */
+
+#include "bgpstream_bgpdump.h"
+#include "bgpstream_int.h"
+#include "bgpstream_elem_int.h"
+#include "bgpstream_log.h"
+#include <assert.h>
+#include <stdio.h>
+#include <inttypes.h>
+
+#define B_REMAIN (len - written)
+#define B_FULL (written >= len)
+#define ADD_PIPE                                                               \
+  do {                                                                         \
+    if (B_REMAIN > 1) {                                                        \
+      *buf_p = '|';                                                            \
+      buf_p++;                                                                 \
+      *buf_p = '\0';                                                           \
+      written++;                                                               \
+    } else {                                                                   \
+      return NULL;                                                             \
+    }                                                                          \
+  } while (0)
+
+#define SEEK_STR_END                                                           \
+  do {                                                                         \
+    while (*buf_p != '\0') {                                                   \
+      written++;                                                               \
+      buf_p++;                                                                 \
+    }                                                                          \
+  } while (0)
+
+char *bgpstream_record_elem_bgpdump_snprintf(char *buf, size_t len,
+                                     bgpstream_record_t *record,
+                                     bgpstream_elem_t *elem)
+{
+  assert(record);
+  assert(elem);
+
+  size_t written = 0; /* < how many bytes we wanted to write */
+  ssize_t c = 0;      /* < how many chars were written */
+  char *buf_p = buf;
+
+  /* Record type */
+  switch (elem->type) {
+  case BGPSTREAM_ELEM_TYPE_RIB:
+    c = snprintf(buf_p, B_REMAIN, "TABLE_DUMP2|%u", record->time_sec);
+    break;
+  case BGPSTREAM_ELEM_TYPE_ANNOUNCEMENT:
+  case BGPSTREAM_ELEM_TYPE_WITHDRAWAL:
+  case BGPSTREAM_ELEM_TYPE_PEERSTATE:
+    c = snprintf(buf_p, B_REMAIN, "BGP4MP|%u", record->time_sec);
+    break;
+  default:
+    break;
+  }
+  written += c;
+  buf_p += c;
```

Comment:
i _think_ anytime you do this, you need a corresponding `if(B_FULL) return NULL;`

Reply:
looking at other elem snprintf function it seems that it has been missed occasionally there too

Reply:
oh, i take it back. it is only needed when not calling `ADD_PIPE` since that does a length check.